chain-test/src/test/scala/org/bitcoins/chain/util/ChainUnitTest.scala

```diff
@@ -23,77 +39,269 @@ trait ChainUnitTest
 
-  lazy val chainHandler: ChainHandler = ChainHandler(blockchain)
-  lazy val chainApi: ChainApi = chainHandler
+  private lazy val chainHandler: ChainHandler = ChainHandler(blockchain)
 
   implicit def ec: ExecutionContext =
     scala.concurrent.ExecutionContext.Implicits.global
 
-  /** Fixture that creates a [[org.bitcoins.chain.models.BlockHeaderTable]]
+  def makeDependentFixture[T](
+      build: () => Future[T],
+      destroy: T => Future[Any])(test: OneArgAsyncTest): FutureOutcome = {
+    val fixtureF = build()
+
+    val outcomeF = fixtureF.flatMap { fixture =>
+      test(fixture.asInstanceOf[FixtureParam]).toFuture
+    }
+
+    val destroyP = Promise[Unit]()
+    outcomeF.onComplete { _ =>
+      fixtureF.foreach { fixture =>
+        destroy(fixture).onComplete {
+          case Success(_)   => destroyP.success(())
+          case Failure(err) => destroyP.failure(err)
+        }
+      }
+    }
+
+    val outcomeAfterDestroyF = destroyP.future.flatMap(_ => outcomeF)
+
+    new FutureOutcome(outcomeAfterDestroyF)
+  }
+
+  def makeFixture[T](build: () => Future[T], destroy: () => Future[Any])(
+      test: OneArgAsyncTest): FutureOutcome = {
+    val outcomeF = build().flatMap { fixture =>
+      test(fixture.asInstanceOf[FixtureParam]).toFuture
+    }
+
+    val destroyP = Promise[Unit]()
+    outcomeF.onComplete { _ =>
+      destroy().onComplete {
+        case Success(_)   => destroyP.success(())
+        case Failure(err) => destroyP.failure(err)
+      }
+    }
+
+    val outcomeAfterDestroyF = destroyP.future.flatMap(_ => outcomeF)
+
+    new FutureOutcome(outcomeAfterDestroyF)
+  }
+
+  def composeBuilders[T, U](
+      builder: () => Future[T],
+      dependentBuilder: T => Future[U]): () => Future[(T, U)] = () => {
+    builder().flatMap { first =>
+      dependentBuilder(first).map { second =>
+        (first, second)
+      }
+    }
+  }
+
+  def composeBuildersAndWrap[T, U, C](
+      builder: () => Future[T],
+      dependentBuilder: T => Future[U],
+      wrap: (T, U) => C): () => Future[C] = () => {
+    composeBuilders(builder, dependentBuilder)().map {
+      case (first, second) => wrap(first, second)
+    }
+  }
+
+  def createBlockHeaderDAO(): Future[BlockHeaderDAO] = {
+    val genesisHeaderF = setupHeaderTableWithGenesisHeader()
+
+    genesisHeaderF.map(_ => blockHeaderDAO)
+  }
+
+  def destroyHeaderTable(): Future[Unit] = {
+    ChainDbManagement.dropHeaderTable(dbConfig)
+  }
+
+  /**
+    * Fixture that creates a [[org.bitcoins.chain.models.BlockHeaderTable]]
     * with one row inserted into it, the [[RegTestNetChainParams]]
     * genesis block
-    * @param test
-    * @return
     */
   def withBlockHeaderDAO(test: OneArgAsyncTest): FutureOutcome = {
+    makeFixture(createBlockHeaderDAO, destroyHeaderTable)(test)
+  }
 
-    val genesisHeaderF = setupHeaderTableWithGenesisHeader()
+  def createPopulatedBlockHeaderDAO(): Future[BlockHeaderDAO] = {
```

Comment:
Add comment explaining that we are reading stored headers between heights X and Y and inserting them into a table.